authz/grpc_audit_end2end_test.go

```diff
@@ -0,0 +1,304 @@
+package authz
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"io"
+	"net"
+	"os"
+	"testing"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/authz/audit"
+	"google.golang.org/grpc/credentials"
+	testgrpc "google.golang.org/grpc/interop/grpc_testing"
+	testpb "google.golang.org/grpc/interop/grpc_testing"
+	"google.golang.org/grpc/testdata"
+
+	_ "google.golang.org/grpc/authz/audit/stdout"
+)
+
+type testServer struct {
+	testgrpc.UnimplementedTestServiceServer
+}
+
+func (s *testServer) UnaryCall(ctx context.Context, req *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
+	return &testpb.SimpleResponse{}, nil
+}
+
+type statAuditLogger struct {
+	Stat         map[bool]int
+	stdoutLogger audit.Logger
+	SpiffeIds    []string
+}
+
+func (s *statAuditLogger) Log(event *audit.Event) {
+	s.stdoutLogger.Log(event)
+	if event.Authorized {
+		s.Stat[true]++
+	} else {
+		s.Stat[false]++
+	}
+	s.SpiffeIds = append(s.SpiffeIds, event.Principal)
+}
+
+type loggerBuilder struct {
+	Stat      map[bool]int
+	SpiffeIds []string
+}
+
+func (loggerBuilder) Name() string {
+	return "stat_logger"
+}
+func (lb *loggerBuilder) Build(audit.LoggerConfig) audit.Logger {
+	return &statAuditLogger{
+		Stat:         lb.Stat,
```

Comment:
Do we need to pass through pointers/addresses here?

Reply:
IIUC we don't need but it's a kind of good default. @rockspore WDYT?

Reply:
Go maps are passed by references already. We don't need to worry about it being copied here.